client/tests/gui/test_shortcuts.py

```diff
@@ -0,0 +1,24 @@
+import re
+from collections import defaultdict
+from pathlib import Path
+
+
+def test_shortcuts_are_unique(mocker):
+    """
+    Look through the codebase for setShortcut() calls,
+    verify the argument passed is unique
```

Comment:
I didn't go into this level of detail in <https://github.com/freedomofpress/securedrop-client/wiki/Keyboard-Shortcuts>, but it's worth noting that this is not a strict requirement, just a policy we could adopt (and use this test to help enforce).  Here's what I've realized after sleeping on it:

Shortcuts don't have to be unique *globally* *always*; they have to be unique *[per context](https://doc.qt.io/qt-6/qt.html#ShortcutContext-enum)* *per point in time*.  So this is a valid counterexample across times:

| Time | Context | Shortcut | Action |
| --- | --- | --- | --- |
| 1 | `WindowShortcut` (default) | <kbd>Ctrl</kbd> + <kbd>Q</kbd> | Quit |
| 2 | `WindowShortcut` (default) | <kbd>Ctrl</kbd> + <kbd>Q</kbd> | File bug report |

Now, this would be a terrible UX, which this test helps prevent.  But here's another valid (and more realistic) counterexample across contexts and widgets:

| Time | Context | Shortcut | Action |
| --- | --- | --- | --- |
| 1 | `WidgetShortcut` for `ReplyBox` 1 | <kbd>Ctrl</kbd> + <kbd>Enter</kbd> | Send from `ReplyBox` 1 |
| 1 | `WidgetShortcut` for `ReplyBox` 2 | <kbd>Ctrl</kbd> + <kbd>Enter</kbd> | Send from `ReplyBox` 2[^1] |

Again, this is valid!  The shortcuts don't collide *in the UI* because they're scoped to the right context.  But they also don't collide *in the code*, and wouldn't be caught by this test, because they're only defined once, on the same widget class before it's instantiated multiple times.  With the `WindowShortcut` context, they *would* collide, and Qt would complain about them, but they still wouldn't be caught by this test.

So a rigorous test would have to search for duplicate `(shortcut, context)` pairs[^2] for all combinations of widgets actually displayed in a program run.  That's at best a state-space explosion and at worst the halting problem: either way, infeasible.

In conclusion: I still think this test is a good linting measure; it's just both (a) not a complete check and (b) quite nuanced in what it *does* check.  So let's just document that:

```suggestion
    verify the argument passed is unique.  This won't catch collisions between instances of a class that calls `setShortcut()` without also adjusting `setShortcutContext()`.  It also enforces a policy that we don't want to reuse the same global shortcut for different purposes (or even with different handlers) at different times.  Either of these limitations may be worth revisiting for UX reasons in the future.
```

[^1]: In some application that displays multiple instances of the same widget at one time.  Which we do, just not `ReplyBox`es.

[^2]: Actually duplicate shortcuts down the [`ShortcutContext`](https://doc.qt.io/qt-6/qt.html#ShortcutContext-enum) hierarchy, but I digress....

Reply:
> So a rigorous test would have to search for duplicate `(shortcut, context)` pairs[2](#user-content-fn-2-028b8555d71dc093ef9160a985efd927) for all combinations of widgets actually displayed in a program run. That's at best a state-space explosion and at worst the halting problem: either way, infeasible.

Yeah. I wonder if we can add a hook to `QAction::event: Ambiguous shortcut overload` that causes tests to fail, but from what I can tell, it only actual triggers when you press the shortcut, so we'd need tests that actually test the shortcuts...